endpoints/bookstore-grpc/client/src/main/java/com/google/endpoints/examples/bookstore/BookstoreClient.java

```diff
@@ -0,0 +1,238 @@
+      printUsage(options);
+      return;
+    }
+
+    String address = params.getOptionValue("bookstore", DEFAULT_ADDRESS);
+    String apiKey = params.getOptionValue("api_key");
+    String authToken = params.getOptionValue("auth_token");
+    String operation = params.getOptionValue("operation", "list");
+
+    // Create gRPC stub.
+    BookstoreGrpc.BookstoreBlockingStub bookstore = createBookstoreStub(
+        address, apiKey, authToken);
+
+    if ("list".equals(operation)) {
+      listShelves(bookstore);
+    } else if ("create".equals(operation)) {
+      createShelf(bookstore);
+    } else if ("enumerate".equals(operation)) {
+      enumerate(bookstore);
+    }
+  }
+
+  /**
+   * Lists all shelves in the bookstore.
+   *
+   * @param bookstore a client stub to call Bookstore service.
+   */
+  static void listShelves(BookstoreGrpc.BookstoreBlockingStub bookstore) {
+    ListShelvesResponse shelves = bookstore.listShelves(Empty.getDefaultInstance());
+    System.out.println(shelves);
+  }
+
+  /**
+   * Creates a new shelf in the bookstore.
+   *
+   * @param bookstore a client stub to call Bookstore service.
+   */
+  static void createShelf(BookstoreGrpc.BookstoreBlockingStub bookstore) {
+    CreateShelfRequest.Builder builder = CreateShelfRequest.newBuilder();
+    builder.getShelfBuilder().setTheme("Computers");
+    Shelf shelf = bookstore.createShelf(builder.build());
+    System.out.println(shelf);
+  }
+
+  /**
+   * Enumerates all books in the bookstore.
+   *
+   * @param bookstore a client stub to call Bookstore service.
+   */
+  static void enumerate(BookstoreGrpc.BookstoreBlockingStub bookstore) {
+    System.out.println("Calling listShelves");
+    ListShelvesResponse shelves = bookstore.listShelves(Empty.getDefaultInstance());
+    System.out.println(shelves);
+
+    for (Shelf s : shelves.getShelvesList()) {
+      System.out.format("Getting shelf %d\n", s.getId());
+      GetShelfRequest getShelfRequest = GetShelfRequest.newBuilder()
+          .setShelf(s.getId())
+          .build();
+      Shelf shelf = bookstore.getShelf(getShelfRequest);
+      System.out.println(shelf);
+
+      System.out.format("Getting books from shelf %d:\n", shelf.getId());
+      ListBooksRequest listBooksRequest = ListBooksRequest.newBuilder()
+          .setShelf(shelf.getId())
+          .build();
+
+      ListBooksResponse books = bookstore.listBooks(listBooksRequest);
+      System.out.println(books);
+
+      for (Book b : books.getBooksList()) {
+        System.out.format("Getting book %d from shelf %d:\n", b.getId(), shelf.getId());
+
+        GetBookRequest getBookRequest = GetBookRequest.newBuilder()
+            .setShelf(shelf.getId())
+            .setBook(b.getId())
+            .build();
+
+        Book book = bookstore.getBook(getBookRequest);
+        System.out.println(book);
+      }
+    }
+  }
+
+  private static final class Interceptor implements ClientInterceptor {
```

Comment:
A javadoc describing what this class is for would probably also be helpful, since it's doing something that's not necessarily intuitive.